pkg/k8s/k8s_service.go

```diff
@@ -142,3 +144,92 @@ func prepareService(
 	}
 	return svc, nil
 }
+
+func (c *Client) WaitForService(ctx context.Context, name string) error {
+	for {
+		service, err := c.clientset.CoreV1().Services(c.namespace).Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			return ErrGettingService.WithParams(name).Wrap(err)
+		}
+
+		if service.Spec.Type == v1.ServiceTypeLoadBalancer {
+			if len(service.Status.LoadBalancer.Ingress) == 0 {
+				time.Sleep(waitRetry)
+				continue // Wait until the LoadBalancer IP is available
+			}
+		} else if service.Spec.Type == v1.ServiceTypeNodePort {
+			if service.Spec.Ports[0].NodePort == 0 {
+				return ErrNodePortNotSet
+			}
+		} else if len(service.Spec.ExternalIPs) == 0 {
+			return ErrExternalIPsNotSet
+		}
+
+		// Check if service is reachable
+		endpoint, err := c.GetServiceEndpoint(ctx, name)
+		if err != nil {
+			return ErrGettingServiceEndpoint.WithParams(name).Wrap(err)
+		}
+
+		if err := checkServiceConnectivity(endpoint); err != nil {
+			time.Sleep(waitRetry) // Retry after some seconds if the service is not reachable
+			continue
+		}
+
+		break // Service is reachable, exit the loop
+	}
+
+	select {
+	case <-ctx.Done():
+		return ErrTimeoutWaitingForServiceReady
+	default:
+		return nil
+	}
+}
+
+func (c *Client) GetServiceEndpoint(ctx context.Context, name string) (string, error) {
+	srv, err := c.clientset.CoreV1().Services(c.namespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return "", ErrGettingService.WithParams(name).Wrap(err)
+	}
+
+	if srv.Spec.Type == v1.ServiceTypeLoadBalancer {
```

Comment:
Similar comment about using a switch statement to understand what happens if it is not one of these two?

Reply:
I think that would also make it clearer how the return values are being handles as currently the return statement is for this if condition at the bottom, which is kind of far away from the context of the return values.

Reply:
Thanks for pointing it out. Ill fix it